### DISCUSSION TITLE
OAuth 2.0 인증 과정

### DISCUSSION THREAD

**Discussion by jaejae-yoo** [woowacourse-teams/2022-moamoa#36] [공유]
OAuth는 인증을 위한 개방형 표준 프로토콜로, 사용자가 인터넷 서비스(facebook, 카카오톡 등) 기능을 다른 애플리케이션에서도 사용할 수 있도록 한 것입니다.

## OAuth 인증 과정
<img width="1046" alt="스크린샷 2022-07-10 오후 1 22 36" src="https://user-images.githubusercontent.com/61091307/178145694-0f3bac80-25c3-40e8-9707-6205d9dbfce0.png">

## OAuth Flow
<img width="1210" alt="스크린샷 2022-07-10 오후 1 25 26" src="https://user-images.githubusercontent.com/61091307/178145700-83a3b37e-f3c2-4978-97c8-7e6f58c04c48.png">

1) FE가 Github에 ID와 비밀번호를 가지고 로그인을 요청합니다. (이때 client_id와 redirect_uri를 파라미터로 같이 넘깁니다.)

```
GET https://github.com/login/oauth/authorize

Parameters:
client_id, redirect_uri
```
2) Github는 FE에게 redirect_uri로 **Authorization code**를 줍니다.

3) FE는 Github에게 받은 Authorization code를 BE에게 줍니다.

4) BE는 Authorization code를 사용해서, access_token을 요청합니다.

```
POST https://github.com/login/oauth/access_token

Parameters:
client_id, client_secret, code
```
5) BE는 Github으로 부터 **access_token**을 받습니다.

6) BE는 받은 access_token을 사용해서 필요한 리소스(ex. 깃허브 프로필 정보)를 요구합니다.

```
GET https://api.github.com/user

Authorization:
access_token
```
7) Github는 BE에 리소스를 제공합니다.

8) **받은 리소스의 정보를 활용해서 FE와 통신할 token을 발급**합니다.

9) FE는 발급받은 token을 사용해서 BE와 통신합니다.

**Comment by tco0427** [comment]
User, Clinet(MOAMOA), Authorization, Resource 측면과 FE, BE, Github 두 가지 관점에서 그림을 제공해줘서 이해가 훨씬 수월하네요..!!
밑에 부가 설명도 이해하는데 도움이 되었습니다!!
